### ISSUE TITLE
exec node doesn't handle whitespace in command path (+ suggested fix)

### ISSUE THREAD

**Issue by hreimer** [node-red/node-red#1100]
When setting a path with whitespace as the command for the exec node, it is treated as a whitespace-separated argument by the exec function.

System: 
Node-RED version: v0.15.2
Node.js  version: v5.12.0
Mac OS X 10.11.6 | Darwin 15.6.0 x64 LE

I posted this question already on [stackoverflow](http://stackoverflow.com/questions/41475554/how-to-escape-whitespace-in-node-red-exec-node-command), and was wondering if there is another solution to use commands with whitespace with the builtin exec node, or if should just resort with using my customised exec node, or if you would integrate my fix into the repository (just trying to contribute, as a pull request didn't seem reasonable for such a small fix), lines 46-52 in the [nodes/core/core/75-exec.js](https://github.com/node-red/node-red/blob/master/nodes/core/core/75-exec.js) file that I cannot attach: 

                var arg = "";
                if ((node.addpay === true) && msg.hasOwnProperty("payload")) { arg += " "+msg.payload; }
                if (node.append.trim() !== "") { arg += " "+node.append; }
                // slice whole line by spaces (trying to honour quotes);
                arg = arg.match(/(?:[^\s"]+|"[^"]*")+/g);
                arg.unshift(node.cmd);

Basically the fix was to prepend the command string to the arguments array after slicing instead of appending the arguments to the command string and then slice it into an array by the whitespace separator.

If you agree this is a helpful and correct fix it would be great if you could integrate it - this is my first time contributing to a repository, hope I did it the right way. If not I appreciate any hints how to improve.

**Comment by dceejay** [comment]
can you not just quote the string ?
if I pass in  "/home/pi/foo bar" to an exec node set to ls - I get the expected result ?
Will this fix not break existing behaviour if trying to pass in multiple arguments ?

**Comment by dceejay** [comment]
I also tried copying a binary to directory "foo bar" and can also execute that fine - eg on my pi
 
"/home/pi/foo bar/date"    works OK

**Comment by hreimer** [comment]
Quoting the string does not work..as well as escaping the whitespace with `\\`or `^` or any other character.

did you use `ls -l` as the command in the exec node? If so, then the `-l`would be extracted as the first argument after the whitespace, which is fine (in this particular case).

If I try the same as you, copying `/bin/date` into a custom `/Users/<username>/foo bar/` directory it still doesn't work and as soon as the whitespace is replaced with any other character, it works.

Maybe this is an issue with OS X when it works on your pi?

I also didn't have any problems with multiple arguments, as I use a function node with multiple arguments separated by whitespace (e.g. `msg.payload = "--saveasimg --imgformat=PNG --resolution=800x600 " + pdfFilePath;`where pdfFilePath is a variable) and append them to the exec node, which is basically `/opt/path with whitespace/executable`. A console.log in the 75-exec.js file shows that all (multiple) arguments are copied exactly as before, but instead of splitting the command path into arguments (making`/opt/path ` the first argument which then becomes the invalid command) because of the whitespace it is now keeping the original command path and prepending it after the split, therefore resulting in a valid `/opt path with whitespace/executable` string.

(Hope I could explain myself correctly, English is not my first language)

**Comment by dceejay** [comment]
just tried on OSX and it's fine there as well

maybe share your complete flow (or the relevant three nodes.

Here's my trivial flow

[{"id":"4dc74222.1c66bc","type":"inject","z":"6e61872d.919e78","name":"","topic":"","payload":"-R","payloadType":"str","repeat":"","crontab":"","once":false,"x":172.5,"y":343,"wires":[["9f6da703.49f408"]]},{"id":"9f6da703.49f408","type":"exec","z":"6e61872d.919e78","command":"\"/home/pi/foo bar/date\"","addpay":true,"append":"","useSpawn":"","timer":"","name":"","x":389.5,"y":347.5,"wires":[["644bb9df.6f31a8"],[],[]]},{"id":"644bb9df.6f31a8","type":"debug","z":"6e61872d.919e78","name":"","active":true,"console":"false","complete":"false","x":656.5,"y":338,"wires":[]}]

**Comment by hreimer** [comment]
thanks for your flow - I tried it with yours which is almost the same as my minimal flow is, and I think I found the cause of the problem: the path with whitespace is quoted just as yours, but when I check the "use spawn() instead of exec()?" Option (on either flow), it returns -2 and thats where my problem is. Leave that option unchecked and it works, but since I need spawn() for my program to work and and the exec node should work with spawn and exec, that's not really a solution.. can you reproduce it?

my minimal flow:

[{"id":"b6633d56.4ed2a","type":"tab","label":"Flow 2"},{"id":"4f1057b3.769ee8","type":"inject","z":"b6633d56.4ed2a","name":"","topic":"","payload":"-r 1000","payloadType":"str","repeat":"","crontab":"","once":false,"x":190,"y":120,"wires":[["38db37c5.221c28"]]},{"id":"38db37c5.221c28","type":"exec","z":"b6633d56.4ed2a","command":"\"/opt/foo bar/date\"","addpay":true,"append":"","useSpawn":true,"timer":"","name":"","x":397,"y":124.5,"wires":[["df9d7dee.f547e"],[],[]]},{"id":"df9d7dee.f547e","type":"debug","z":"b6633d56.4ed2a","name":"","active":true,"console":"false","complete":"false","x":674,"y":115,"wires":[]}]

**Comment by dceejay** [comment]
Aha - right - on the same page now...
and this just got slightly trickier due to the way spawn doesn't like spaces.
We can't just fix it by adding the command part last because some people (myself included) also add parts of parameters to that section.. - eg in your example the -t "could" be part of the command and we just pass in the number...  so... my proposal is
that we use quotes (as works with exec) to surround the command part with spaces but remove them from the command before passing .... 

In the original code this would be adding a line 52
       51         var cmd = arg.shift();
       52         if (/^".*"$/.test(cmd)) { cmd = cmd.slice(1,-1); }

Thoughts ?

**Comment by dceejay** [comment]
added to 0.16 branch - 08b11addecbfb0da7b3e0efce7ae0c87987cc981

**Comment by knolleary** [comment]
@dceejay following your example with the `-t`, does your fix allow:

    "foo bar" -t

because that regex looks like it expects the quotes to wrap the entire property.

**Comment by dceejay** [comment]
yeah - that's after a split that honours quotes - then the shift takes the first chunk (the one with quotes)

**Comment by hreimer** [comment]
@dceejay you're right, I didn't think about the additional parameter input in the edit panel of the exec node or directly in the command part.

Your solution works for me with all possible combinations, thanks for that!
